File: app/ui/dialogs/batch_dialog.py
import os
from PySide6.QtCore import Qt
from PySide6.QtWidgets import (QDialog, QVBoxLayout, QHBoxLayout, QLabel,
                             QComboBox, QPushButton, QLineEdit, QFileDialog, QProgressBar, QListWidget, QGroupBox)
from app.workers.batch_worker import BatchWorker
from app.utils.logger import logger

class BatchDialog(QDialog):
    """Dialog window for batch background removal across multiple files/folders."""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Batch Background Removal")
        self.setFixedSize(560, 480)

        self.file_paths = []
        self.worker = None

        layout = QVBoxLayout(self)
        layout.setContentsMargins(16, 16, 16, 16)
        layout.setSpacing(12)

        # Input File Queue
        grp_in = QGroupBox("Input Images Queue", self)
        vbox_in = QVBoxLayout(grp_in)

        self.list_files = QListWidget(self)
        vbox_in.addWidget(self.list_files)

        hbox_in_btns = QHBoxLayout()
        btn_add_files = QPushButton("➕ Add Files...", self)
        btn_add_files.clicked.connect(self._add_files)
        hbox_in_btns.addWidget(btn_add_files)

        btn_add_dir = QPushButton("📁 Add Folder...", self)
        btn_add_dir.clicked.connect(self._add_folder)
        hbox_in_btns.addWidget(btn_add_dir)

        btn_clear = QPushButton("🗑️ Clear Queue", self)
        btn_clear.clicked.connect(self._clear_queue)
        hbox_in_btns.addWidget(btn_clear)
        vbox_in.addLayout(hbox_in_btns)

        layout.addWidget(grp_in)

        # Output Folder & Options
        grp_out = QGroupBox("Batch Settings", self)
        vbox_out = QVBoxLayout(grp_out)

        hbox_out_dir = QHBoxLayout()
        hbox_out_dir.addWidget(QLabel("Output Dir:", self))
        self.txt_out_dir = QLineEdit(self)
        self.txt_out_dir.setText(os.path.join(os.path.expanduser("~"), "BackgroundRemover_Output"))
        hbox_out_dir.addWidget(self.txt_out_dir)

        btn_browse_out = QPushButton("Browse...", self)
        btn_browse_out.clicked.connect(self._browse_output)
        hbox_out_dir.addWidget(btn_browse_out)
        vbox_out.addLayout(hbox_out_dir)

        hbox_opts = QHBoxLayout()
        hbox_opts.addWidget(QLabel("AI Model:", self))
        self.combo_model = QComboBox(self)
        self.combo_model.addItems(["RMBG-1.4", "U2Net", "Silueta"])
        hbox_opts.addWidget(self.combo_model)

        hbox_opts.addWidget(QLabel("Format:", self))
        self.combo_fmt = QComboBox(self)
        self.combo_fmt.addItems(["PNG", "JPG", "WEBP"])
        hbox_opts.addWidget(self.combo_fmt)
        vbox_out.addLayout(hbox_opts)

        layout.addWidget(grp_out)

        # Progress Section
        self.progress_bar = QProgressBar(self)
        self.progress_bar.setValue(0)
        layout.addWidget(self.progress_bar)

        self.lbl_status = QLabel("Ready to process.", self)
        layout.addWidget(self.lbl_status)

        # Action Buttons
        hbox_actions = QHBoxLayout()
        hbox_actions.addStretch()
        self.btn_start = QPushButton("⚡ Start Batch", self)
        self.btn_start.setObjectName("btn_primary")
        self.btn_start.clicked.connect(self._start_batch)
        hbox_actions.addWidget(self.btn_start)

        btn_close = QPushButton("Close", self)
        btn_close.clicked.connect(self.accept)
        hbox_actions.addWidget(btn_close)
        layout.addLayout(hbox_actions)

    def _add_files(self):
        files, _ = QFileDialog.getOpenFileNames(self, "Select Images", "", "Images (*.png *.jpg *.jpeg *.webp *.bmp *.tiff)")
        for f in files:
            if f not in self.file_paths:
                self.file_paths.append(f)
                self.list_files.addItem(os.path.basename(f))
        self.lbl_status.setText(f"{len(self.file_paths)} files queued.")

    def _add_folder(self):
        folder = QFileDialog.getExistingDirectory(self, "Select Image Directory")
        if folder:
            valid_exts = (".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tiff")
            count = 0
            for root, _, files in os.walk(folder):
                for file in files:
                    if file.lower().endswith(valid_exts):
                        full_p = os.path.join(root, file)
                        if full_p not in self.file_paths:
                            self.file_paths.append(full_p)
                            self.list_files.addItem(file)
                            count += 1
            self.lbl_status.setText(f"Added {count} files. Total queued: {len(self.file_paths)}")

    def _clear_queue(self):
        self.file_paths.clear()
        self.list_files.clear()
        self.lbl_status.setText("Queue cleared.")
        self.progress_bar.setValue(0)

    def _browse_output(self):
        folder = QFileDialog.getExistingDirectory(self, "Select Output Directory")
        if folder:
            self.txt_out_dir.setText(folder)

    def _start_batch(self):
        if not self.file_paths:
            self.lbl_status.setText("Error: Queue is empty.")
            return

        out_dir = self.txt_out_dir.text()
        model_name = self.combo_model.currentText()
        fmt = self.combo_fmt.currentText()

        self.btn_start.setEnabled(False)
        self.worker = BatchWorker(self.file_paths, out_dir, model_name=model_name, output_format=fmt)
        self.worker.progress.connect(self._on_progress)
        self.worker.finished.connect(self._on_finished)
        self.worker.start()

    def _on_progress(self, idx: int, total: int, filename: str):
        pct = int((idx / float(total)) * 100)
        self.progress_bar.setValue(pct)
        self.lbl_status.setText(f"Processing ({idx}/{total}): {filename}")

    def _on_finished(self, success: int, fail: int):
        self.btn_start.setEnabled(True)
        self.lbl_status.setText(f"Batch completed! Success: {success}, Failed: {fail}")

File: Cpp/src/ui/dialogs/BatchDialog.cpp
#include "ui/dialogs/BatchDialog.h"
#include "workers/BatchWorker.h"
#include "utils/Logger.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGroupBox>
#include <QFileDialog>
#include <QDir>
#include <QDirIterator>

namespace ImageCut {
namespace UI {

BatchDialog::BatchDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Batch Background Removal");
    setFixedSize(560, 480);

    QVBoxLayout* layout = new QVBoxLayout(this);
    layout->setContentsMargins(16, 16, 16, 16);
    layout->setSpacing(12);

    QGroupBox* grpIn = new QGroupBox("Input Images Queue", this);
    QVBoxLayout* vboxIn = new QVBoxLayout(grpIn);
    m_listFiles = new QListWidget(this);
    vboxIn->addWidget(m_listFiles);

    QHBoxLayout* hboxInBtns = new QHBoxLayout();
    QPushButton* btnAddFiles = new QPushButton("➕ Add Files...", this);
    connect(btnAddFiles, &QPushButton::clicked, this, &BatchDialog::addFiles);
    hboxInBtns->addWidget(btnAddFiles);

    QPushButton* btnAddDir = new QPushButton("📁 Add Folder...", this);
    connect(btnAddDir, &QPushButton::clicked, this, &BatchDialog::addFolder);
    hboxInBtns->addWidget(btnAddDir);

    QPushButton* btnClear = new QPushButton("🗑️ Clear Queue", this);
    connect(btnClear, &QPushButton::clicked, this, &BatchDialog::clearQueue);
    hboxInBtns->addWidget(btnClear);
    vboxIn->addLayout(hboxInBtns);
    layout->addWidget(grpIn);

    QGroupBox* grpOut = new QGroupBox("Batch Settings", this);
    QVBoxLayout* vboxOut = new QVBoxLayout(grpOut);

    QHBoxLayout* hboxOutDir = new QHBoxLayout();
    hboxOutDir->addWidget(new QLabel("Output Dir:", this));
    m_txtOutDir = new QLineEdit(this);
    m_txtOutDir->setText(QDir::homePath() + "/BackgroundRemover_Output");
    hboxOutDir->addWidget(m_txtOutDir);

    QPushButton* btnBrowseOut = new QPushButton("Browse...", this);
    connect(btnBrowseOut, &QPushButton::clicked, this, &BatchDialog::browseOutput);
    hboxOutDir->addWidget(btnBrowseOut);
    vboxOut->addLayout(hboxOutDir);

    QHBoxLayout* hboxOpts = new QHBoxLayout();
    hboxOpts->addWidget(new QLabel("AI Model:", this));
    m_comboModel = new QComboBox(this);
    m_comboModel->addItems({ "RMBG-1.4", "U2Net", "Silueta" });
    hboxOpts->addWidget(m_comboModel);

    hboxOpts->addWidget(new QLabel("Format:", this));
    m_comboFmt = new QComboBox(this);
    m_comboFmt->addItems({ "PNG", "JPG", "WEBP" });
    hboxOpts->addWidget(m_comboFmt);
    vboxOut->addLayout(hboxOpts);

    layout->addWidget(grpOut);

    m_progressBar = new QProgressBar(this);
    m_progressBar->setValue(0);
    layout->addWidget(m_progressBar);

    m_lblStatus = new QLabel("Ready to process.", this);
    layout->addWidget(m_lblStatus);

    QHBoxLayout* hboxActions = new QHBoxLayout();
    hboxActions->addStretch();
    m_btnStart = new QPushButton("⚡ Start Batch", this);
    m_btnStart->setObjectName("btn_primary");
    connect(m_btnStart, &QPushButton::clicked, this, &BatchDialog::startBatch);
    hboxActions->addWidget(m_btnStart);

    QPushButton* btnClose = new QPushButton("Close", this);
    connect(btnClose, &QPushButton::clicked, this, &QDialog::accept);
    hboxActions->addWidget(btnClose);
    layout->addLayout(hboxActions);
}

void BatchDialog::addFiles() {
    QStringList files = QFileDialog::getOpenFileNames(this, "Select Images", "", "Images (*.png *.jpg *.jpeg *.webp *.bmp *.tiff)");
    for (const auto& f : files) {
        if (!m_filePaths.contains(f)) {
            m_filePaths.append(f);
            QFileInfo info(f);
            m_listFiles->addItem(info.fileName());
        }
    }
    m_lblStatus->setText(QString("%1 files queued.").arg(m_filePaths.size()));
}

void BatchDialog::addFolder() {
    QString folder = QFileDialog::getExistingDirectory(this, "Select Image Directory");
    if (!folder.isEmpty()) {
        int count = 0;
        QDirIterator it(folder, QStringList() << "*.png" << "*.jpg" << "*.jpeg" << "*.webp" << "*.bmp" << "*.tiff", QDir::Files, QDirIterator::Subdirectories);
        while (it.hasNext()) {
            QString fullP = it.next();
            if (!m_filePaths.contains(fullP)) {
                m_filePaths.append(fullP);
                m_listFiles->addItem(it.fileName());
                count++;
            }
        }
        m_lblStatus->setText(QString("Added %1 files. Total queued: %2").arg(count).arg(m_filePaths.size()));
    }
}

void BatchDialog::clearQueue() {
    m_filePaths.clear();
    m_listFiles->clear();
    m_lblStatus->setText("Queue cleared.");
    m_progressBar->setValue(0);
}

void BatchDialog::browseOutput() {
    QString folder = QFileDialog::getExistingDirectory(this, "Select Output Directory");
    if (!folder.isEmpty()) {
        m_txtOutDir->setText(folder);
    }
}

void BatchDialog::startBatch() {
    if (m_filePaths.isEmpty()) {
        m_lblStatus->setText("Error: Queue is empty.");
        return;
    }

    m_btnStart->setEnabled(false);
    auto worker = new Workers::BatchWorker(m_filePaths, m_txtOutDir->text(), m_comboModel->currentText(), m_comboFmt->currentText());
    connect(worker, &Workers::BatchWorker::progress, [this](int idx, int total, const QString& filename) {
        int pct = static_cast<int>((idx / static_cast<double>(total)) * 100);
        m_progressBar->setValue(pct);
        m_lblStatus->setText(QString("Processing (%1/%2): %3").arg(idx).arg(total).arg(filename));
    });
    connect(worker, &Workers::BatchWorker::finished, [this, worker](int success, int fail) {
        m_btnStart->setEnabled(true);
        m_lblStatus->setText(QString("Batch completed! Success: %1, Failed: %2").arg(success).arg(fail));
        worker->deleteLater();
    });
    worker->start();
}

} // namespace UI
} // namespace ImageCut
